src/Input/Parser.php

```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
+        
+        // Its not variadic, and we have a constant arg:
+        if ($arg->isConstant()) {
+            $this->setValue($argument, $arg->value());
+            return;
+        }
+
+        // Its not variadic, and we have a variadic arg:
+        if ($arg->isVariadic()) {
+            throw new InvalidArgumentException(
+                sprintf("Argument '%s' is not variadic", $argument->name())
+            );
         }
     }
 
     /**
      * Parse an option, emit its event and set value.
      *
-     * @param string      $arg
-     * @param string|null $nextArg
+     * @param Token      $opt
+     * @param Tokenizer  $tokens
      *
-     * @return bool Whether to eat next arg.
+     * @return int Number of extra tokens consumed
      */
-    protected function parseOptions(string $arg, string $nextArg = null): bool
-    {
-        $value = substr($nextArg ?? '', 0, 1) === '-' ? null : $nextArg;
+    protected function parseOptions(Token $opt, Tokenizer $tokens) : int
+    {   
+
+        // Look ahead for next token:
+        $next = $tokens->offset(1);
+
+        // Get the option:
+        $option = $this->optionFor($opt->value());
+        
+        //Consumed: 
+        $consumed = 0;
+
+        // Unknown option handle it:
+        if (!$option) {
+            // Single value just in case the value is a variadic group:
+            $value = $next ? $next->value() : null;
+            $used  = $this->handleUnknown(
+                $opt->value(), 
+                is_array($value) ? $value[0] ?? null : $value
+            );
+            return $used ? ++$consumed : $consumed;
+        }
+
+        // Early out if its just a flag
+        if (!$next) {
+            $this->setValue($option);
+            return $consumed;
+        }
 
-        if (null === $option  = $this->optionFor($arg)) {
-            return $this->handleUnknown($arg, $value);
+        // If option is variadic, and next is constant, 
+        // then we need to collect all the remaining args:
+        if ($option->variadic() && $next->isConstant()) {
+            $tokens->next();
+            while ($tokens->valid()) {
+                
+                $consumed++;
+
+                if ($tokens->current()->isConstant()) {
+                    $this->setValue($option, $tokens->current()->value());
+                } else {
+                    throw new InvalidParameterException(
+                        "Only constants are allowed in variadic arguments"
+                    );
+                }
+                $tokens->next();
+            }
+            return $consumed;
         }
 
-        $this->_lastVariadic = $option->variadic() ? $option->attributeName() : null;
+        // If option is variadic, and next is a variadic group, 
+        // then we need to collect all the nested values:
+        if ($option->variadic() && $next->isVariadic()) {
+
+            foreach ($next->nested as $token) {
+                if ($token->isConstant()) {
```

Comment:
i see many code repetition of this nature:
```
if (token isConstant) { 
  setValue
} else {
  throw ...
}
```

--- should be extracted to some helper (`setConstantOrFail()`) that can be used like this:
```$this->setConstantOrFail($token)`

Reply:
This is because I want to avoid recursion (as much as possible), and as a general rule I normally don't add more abstraction, especially for internal code which probably will not be used anywhere else...
That said I can add more methods if it's a design rule.

Reply:
this loop is convoluted so if the command definitions for options don't contain any variadic option at all, we should have a way to just use older approach of parsing for performance reasons. i think variadic options are only a small percentage of non variadic options.

Reply:
I'm sorry I don't really understand the point.

My reasoning is:
1. The number of tokens we will iterate through will most of the time, be meaningless, I mean it's a CLI arguments parser...
2. I wanted to reduce function calls as much as possible (for performance) and I wanted to avoid unnecessary Iterations/function calls.
3. I chose this pattern because it's simple, bulletproof, and only consumes/loops/passes values if it's absolutely needed.

In this case unless `$option->variadic()` which means it was declared as variadic and I have a specific `$next->isVariadic()` (which means it's a specific variadic group i.e. [ 1 2 3 ]) - no loop will be reached.

If this is the case all the "grouped" arguments will be consumed in one pass.

Please elaborate